engine/render/src/render/render_script.cpp

```diff
@@ -1372,29 +1408,63 @@ namespace dmRender
     int RenderScript_EnableTexture(lua_State* L)
     {
         RenderScriptInstance* i = RenderScriptInstance_Check(L);
-        dmGraphics::HRenderTarget render_target = 0x0;
 
         uint32_t unit = luaL_checknumber(L, 1);
         if (lua_isnumber(L, 2))
         {
-            render_target = (dmGraphics::HRenderTarget) lua_tonumber(L, 2);
-            int buffer_type_value = (int)luaL_checknumber(L, 3);
-            dmGraphics::BufferType buffer_type = (dmGraphics::BufferType) buffer_type_value;
+            dmGraphics::HAssetHandle asset_handle = (dmGraphics::HAssetHandle) lua_tonumber(L, 2);
+            dmGraphics::AssetType asset_type      = dmGraphics::GetAssetType(asset_handle);
+            dmGraphics::HTexture texture          = 0;
 
-            dmGraphics::HTexture texture = dmGraphics::GetRenderTargetTexture(render_target, buffer_type);
+            if (!dmGraphics::IsAssetHandleValid(i->m_RenderContext->m_GraphicsContext, asset_handle))
+            {
+                char buf[128];
+                return luaL_error(L, "Texture handle '%s' is not valid.", AssetHandleToString(asset_handle, buf, sizeof(buf)));
+            }
+
+            if (asset_type == dmGraphics::ASSET_TYPE_RENDER_TARGET)
+            {
+                dmGraphics::BufferType buffer_type = dmGraphics::BUFFER_TYPE_COLOR0_BIT;
+                if (lua_isnumber(L, 3))
+                {
+                    buffer_type = (dmGraphics::BufferType) lua_tonumber(L, 3);
+                }
+
+                texture = dmGraphics::GetRenderTargetTexture(asset_handle, buffer_type);
+
+                if (texture == 0)
+                {
+                    char buf[128];
+                    return luaL_error(L, "Render target '%s' does not have a texture for the specified buffer type.", AssetHandleToString(asset_handle, buf, sizeof(buf)));
+                }
+            }
+            else if (asset_type == dmGraphics::ASSET_TYPE_TEXTURE)
+            {
+                texture = (dmGraphics::HTexture) asset_handle;
+            }
 
+            // Texture can potentially still be zero if it's an attachment texture
             if(texture != 0)
             {
-                if (InsertCommand(i, Command(COMMAND_TYPE_ENABLE_TEXTURE, unit, (uintptr_t)texture)))
+                // JG: We are passing in the opaque handle here and not the asset handle, this is a workaround for 32 bit systems
+                //     because we are using uintptr_t that is defined as a uint32_t on those platforms. We reconstruct the actual handle
+                //     before actual passing in render_command.cpp that processes the command. And alternative would be to use 64b values for the operands?
+                if (InsertCommand(i, Command(COMMAND_TYPE_ENABLE_TEXTURE, unit, (uintptr_t) dmGraphics::GetOpaqueHandle(texture))))
```

Comment:
Another approach would be to go back to storing the type along the asset in the handle containers instead of void* as it is currently.. But for at least vulkan and console(s) it requires an extra pointer, which would be nice to avoid. As a future improvement we might want to think about representing the adapter structures in a more generic way somehow. Instead of having completely separate OpenGLTexture/VulkanTexture/NullTexture structs it would be nice to split and merge their representations. That way we can merge a lot of boilerplate code paths (e.g getting width/height from a texture doesn't need three separate functions in that case) as well as reduce divergence between them.. For now I don't think we need anything more complex to solve this particular issue, but It's worth to think about

Reply:
Ah, good old C! Yes I think that will be the way to go 👍